[CCE] resource opentelekomcloud_cce_node_pool_v3

# Description of the new resource/data-source
Having the node pools for CCE in Terraform would help us a lot.

I started to write the resource but `make test` is failing and - not being into Go that much - cannot find the issue.

My fork is [here](https://github.com/philippeckel/terraform-provider-opentelekomcloud) and the test fails with

```golang
==> Checking that code complies with gofmt requirements...
go test -i $(go list ./... |grep -v 'vendor') || exit 1
echo $(go list ./... |grep -v 'vendor') | \
        xargs -t -n4 go test  -timeout=30s -parallel=4
go test '-timeout=30s' '-parallel=4' github.com/opentelekomcloud/terraform-provider-opentelekomcloud github.com/opentelekomcloud/terraform-provider-opentelekomcloud/opentelekomcloud 
?       github.com/opentelekomcloud/terraform-provider-opentelekomcloud [no test files]
# github.com/opentelekomcloud/terraform-provider-opentelekomcloud/opentelekomcloud [github.com/opentelekomcloud/terraform-provider-opentelekomcloud/opentelekomcloud.test]
opentelekomcloud/resource_opentelekomcloud_cce_node_pool_v3_test.go:140:29: cannot use rName (type string) as type *testing.T in argument to TestAccCCEClusterV3_basic
opentelekomcloud/resource_opentelekomcloud_cce_node_pool_v3_test.go:140:29: TestAccCCEClusterV3_basic(rName) used as value
FAIL    github.com/opentelekomcloud/terraform-provider-opentelekomcloud/opentelekomcloud [build failed]
FAIL
make: *** [GNUmakefile:13: test] Error 123
```

--- a/docs/resources/cce_node_pool_v3.md
+++ b/docs/resources/cce_node_pool_v3.md
@@ -1,0 +1,120 @@
+---
+subcategory: "Cloud Container Engine (CCE)"
+---
+
+# opentelekomcloud_cce_node_pool_v3
+
+Add a node pool to a container cluster.
+
+## Example Usage
+
+```hcl
+variable "cluster_id" {}
+variable "ssh_key" {}
+variable "availability_zone" {}
+
+resource "opentelekomcloud_cce_node_pool_v3" "node_pool_1" {
+  cluster_id         = var.cluster_id
+  name               = "opentelekomcloud-cce-node-pool_test"
+  os                 = "EulerOS 2.5"
+  flavor             = "s2.xlarge.2"
+  initial_node_count = 2
+  availability_zone  = var.availability_zone
+  key_pair           = var.ssh_key
+
+  scale_enable             = true
+  min_node_count           = 2
+  max_node_count           = 9
+  scale_down_cooldown_time = 100
+  priority                 = 1
+
+  root_volume {
+    size       = 40
+    volumetype = "SSD"
+  }
+  data_volumes {
+    size       = 100
+    volumetype = "SSD"
+  }
+}
+```
+
+## Argument Reference
+The following arguments are supported:
+
+* `cluster_id` - (Required) ID of the cluster. Changing this parameter will create a new resource.
+
+* `flavor` - (Required) Specifies the flavor id. Changing this parameter will create a new resource.
+
+* `availability_zone` - (Required) specify the name of the available partition (AZ). Changing this parameter will create a new resource.
+
+* `key_pair` - (Optional) Key pair name when logging in to select the key pair mode. This parameter and password are alternative. Changing this parameter will create a new resource.
+
+* `password` - (Optional) Key pair name when logging in to select the key pair mode. This parameter and password are alternative. Changing this parameter will create a new resource.
+
+* `os` - (Optional) Node OS. Changing this parameter will create a new resource.
+
+  Supported OS depends on kubernetes version of the cluster.
+  * Clusters of Kubernetes `v1.13` or later support `EulerOS 2.5`.
+  * Clusters of Kubernetes `v1.17` or later support `EulerOS 2.5` and `CentOS 7.7`.
+
+* `name` - (Required) Node Pool Name.
+
+* `initial_node_count` - (Required) Initial number of expected nodes in the node pool.
+
+* `subnet_id` - (Optional) The ID of the subnet to which the NIC belongs. Changing this parameter will create a new resource.
+
+* `preinstall` - (Optional) Script required before installation. The input value can be a Base64 encoded string or not.
+  Changing this parameter will create a new resource.
+
+* `postinstall` - (Optional) Script required after installation. The input value can be a Base64 encoded string or not.
+  Changing this parameter will create a new resource.
+
+* `scale_enable` - (Optional) Whether to enable auto scaling. If Autoscaler is enabled, install the autoscaler add-on to use the auto scaling feature.
+
+* `min_node_count` - (Optional) Minimum number of nodes allowed if auto scaling is enabled.
+
+* `max_node_count` - (Optional) Maximum number of nodes allowed if auto scaling is enabled.
+
+* `scale_down_cooldown_time` - (Optional) Interval between two scaling operations, in minutes.
+
+* `server_group_reference` - (Optional) ECS group ID. If this parameter is specified, all nodes in the node pool will be created in this ECS group.
+
+* `priority` - (Optional) Weight of a node pool. A node pool with a higher weight has a higher priority during scaling.
+
+* `user_tags` - (Optional) Tag of a VM, key/value pair format. Changing this parameter will create a new resource.
+
+* `k8s_tags` - (Optional) Tags of a Kubernetes node, key/value pair format. Changing this parameter will create a new resource.
+
+* `taints` - (Optional) Taints to created nodes to configure anti-affinity. Changing this parameter will create a new resource.
+  * `key` - (Required) A key must contain 1 to 63 characters starting with a letter or digit. Only letters, digits, hyphens (-), underscores (_), and periods (.) are allowed. A DNS subdomain name can be used as the prefix of a key.
+  * `value` - (Required) A value must start with a letter or digit and can contain a maximum of 63 characters, including letters, digits, hyphens (-), underscores (_), and periods (.).
+  * `effect` - (Optional) Available options are `NoSchedule`, `PreferNoSchedule`, and `NoExecute`.
+
+* `root_volume` - (Required) It corresponds to the system disk related configuration. Changing this parameter will create a new resource.
+  * `size` - (Required) Disk size in GB.
+  * `volumetype` - (Required) Disk type.
+  * `extend_param` - (Optional) Disk expansion parameters.
+
+* `data_volumes` - (Required) Represents the data disk to be created. Changing this parameter will create a new resource.
+  * `size` - (Required) Disk size in GB.
+  * `volumetype` - (Required) Disk type.
+  * `extend_param` - (Optional) Disk expansion parameters.
+
+## Attributes Reference
+
+All above argument parameters can be exported as attribute parameters along with attribute reference.
+
+* `status` - Node status information.
+
+* `id` - Specifies a resource ID in UUID format.
+
+* `billing_mode ` - Billing mode of a node.
+
+## Timeouts
+
+This resource provides the following timeouts configuration options:
+
+- `create` - Default is 20 minutes.
+
+- `delete` - Default is 20 minutes.

--- a/opentelekomcloud/custom_functions.go
+++ b/opentelekomcloud/custom_functions.go
@@ -69,3 +69,12 @@ func firstOneSet(res map[string]interface{}, k1, k2 string) (interface{}, error)
 	}
 	return v2, nil
 }
+
+func getHashOrEmpty(v interface{}) string {
+	switch v.(type) {
+	case string:
+		return installScriptHashSum(v.(string))
+	default:
+		return ""
+	}
+}

--- a/opentelekomcloud/customize_diffs.go
+++ b/opentelekomcloud/customize_diffs.go
@@ -186,6 +186,27 @@ func validateVPC(argName string) schema.CustomizeDiffFunc {
 	}
 }
 
+func validateSubnet(argName string) schema.CustomizeDiffFunc {
+	return func(d *schema.ResourceDiff, meta interface{}) error {
+		subnetId := d.Get(argName)
+		if subnetId == nil {
+			return fmt.Errorf(argMissingMsg, argName)
+		}
+		if subnetId == "" {
+			return nil
+		}
+		config := meta.(*Config)
+		subnetClient, err := config.networkingV1Client(GetRegion(d, config))
+		if err != nil {
+			return fmt.Errorf("error creating opentelekomcloud CCE Client: %s", err)
+		}
+		if err := subnets.Get(subnetClient, subnetId.(string)).Err; err != nil {
+			return fmt.Errorf("can't find Subnet `%s`: %s", subnetId, err)
+		}
+		return nil
+	}
+}
+
 func multipleCustomizeDiffs(funcs ...schema.CustomizeDiffFunc) schema.CustomizeDiffFunc {
 	return func(d *schema.ResourceDiff, meta interface{}) error {
 		mErr := &multierror.Error{}

--- a/opentelekomcloud/provider.go
+++ b/opentelekomcloud/provider.go
@@ -253,6 +253,7 @@ func Provider() terraform.ResourceProvider {
 			"opentelekomcloud_cce_addon_v3":                       resourceCCEAddonV3(),
 			"opentelekomcloud_cce_cluster_v3":                     resourceCCEClusterV3(),
 			"opentelekomcloud_cce_node_v3":                        resourceCCENodeV3(),
+			"opentelekomcloud_cce_node_pool_v3":                   resourceCCENodePoolV3(),
 			"opentelekomcloud_ces_alarmrule":                      resourceAlarmRule(),
 			"opentelekomcloud_compute_bms_server_v2":              resourceComputeBMSInstanceV2(),
 			"opentelekomcloud_compute_bms_tags_v2":                resourceBMSTagsV2(),

--- a/opentelekomcloud/resource_opentelekomcloud_cce_node_pool_v3.go
+++ b/opentelekomcloud/resource_opentelekomcloud_cce_node_pool_v3.go
@@ -3,6 +3,7 @@ package opentelekomcloud
 import (
 	"fmt"
 	"log"
+	"regexp"
 	"strings"
 	"time"
 
@@ -16,6 +17,12 @@ import (
 	"github.com/opentelekomcloud/gophertelekomcloud/openstack/common/tags"
 )
 
+var (
+	// Cluster pool taint key and value is 1 to 63 characters starting with a letter or digit.
+	// Only letters, digits, hyphens (-), underscores (_), and periods (.) are allowed.
+	clusterPoolTaintRegex, _ = regexp.Compile("^[a-zA-Z0-9_.-]{1,63}$")
+)
+
 func resourceCCENodePoolV3() *schema.Resource {
 	return &schema.Resource{
 		Create: resourceCCENodePoolV3Create,
@@ -26,7 +33,14 @@ func resourceCCENodePoolV3() *schema.Resource {
 		Timeouts: &schema.ResourceTimeout{
 			Create: schema.DefaultTimeout(20 * time.Minute),
 			Delete: schema.DefaultTimeout(20 * time.Minute),
+			Update: schema.DefaultTimeout(20 * time.Minute),
 		},
+
+		CustomizeDiff: multipleCustomizeDiffs(
+			validateVolumeType("root_volume.*.volumetype"),
+			validateVolumeType("data_volumes.*.volumetype"),
+			validateSubnet("subnet_id"),
+		),
 
 		Schema: map[string]*schema.Schema{
 			"name": {
@@ -69,9 +83,6 @@ func resourceCCENodePoolV3() *schema.Resource {
 						"volumetype": {
 							Type:     schema.TypeString,
 							Required: true,
-							ValidateFunc: validation.StringInSlice([]string{
-								"SATA", "SAS", "SSD",
-							}, true),
 						},
 						"extend_param": {
 							Type:     schema.TypeString,
@@ -93,9 +104,6 @@ func resourceCCENodePoolV3() *schema.Resource {
 						"volumetype": {
 							Type:     schema.TypeString,
 							Required: true,
-							ValidateFunc: validation.StringInSlice([]string{
-								"SATA", "SAS", "SSD", "co-p1", "uh-l1",
-							}, true),
 						},
 						"extend_param": {
 							Type:     schema.TypeString,
@@ -130,14 +138,23 @@ func resourceCCENodePoolV3() *schema.Resource {
 						"key": {
 							Type:     schema.TypeString,
 							Required: true,
+							ValidateFunc: validation.StringMatch(clusterPoolTaintRegex, "Invalid key. "+
+								"Cluster pool taint key is 1 to 63 characters starting with a letter or digit. "+
+								"Only lowercase letters, digits, and hyphens (-) are allowed."),
 						},
 						"value": {
 							Type:     schema.TypeString,
 							Required: true,
+							ValidateFunc: validation.StringMatch(clusterPoolTaintRegex, "Invalid value. "+
+								"Cluster pool taint value is 1 to 63 characters starting with a letter or digit. "+
+								"Only letters, digits, hyphens (-), underscores (_), and periods (.) are allowed."),
 						},
 						"effect": {
 							Type:     schema.TypeString,
 							Required: true,
+							ValidateFunc: validation.StringInSlice([]string{
+								"NoSchedule", "PreferNoSchedule", "NoExecute",
+							}, false),
 						},
 					}},
 			},
@@ -160,30 +177,16 @@ func resourceCCENodePoolV3() *schema.Resource {
 				ForceNew: true,
 			},
 			"preinstall": {
-				Type:     schema.TypeString,
-				Optional: true,
-				ForceNew: true,
-				StateFunc: func(v interface{}) string {
-					switch v.(type) {
-					case string:
-						return installScriptHashSum(v.(string))
-					default:
-						return ""
-					}
-				},
+				Type:      schema.TypeString,
+				Optional:  true,
+				ForceNew:  true,
+				StateFunc: getHashOrEmpty,
 			},
 			"postinstall": {
-				Type:     schema.TypeString,
-				Optional: true,
-				ForceNew: true,
-				StateFunc: func(v interface{}) string {
-					switch v.(type) {
-					case string:
-						return installScriptHashSum(v.(string))
-					default:
-						return ""
-					}
-				},
+				Type:      schema.TypeString,
+				Optional:  true,
+				ForceNew:  true,
+				StateFunc: getHashOrEmpty,
 			},
 			"scale_enable": {
 				Type:     schema.TypeBool,
@@ -308,11 +311,11 @@ func resourceCCENodePoolV3Create(d *schema.ResourceData, meta interface{}) error
 		if _, ok := err.(golangsdk.ErrDefault403); ok {
 			retryNode, err := recursiveNodePoolCreate(nodePoolClient, createOpts, clusterId, 403)
 			if err == "fail" {
-				return fmt.Errorf("error creating Open Telekom Cloud Node Pool")
+				return fmt.Errorf("error creating Open Telekom Cloud CCE Node Pool")
 			}
 			s = retryNode
 		} else {
-			return fmt.Errorf("error creating Open Telekom Cloud Node Pool: %s", err)
+			return fmt.Errorf("error creating Open Telekom Cloud CCE Node Pool: %s", err)
 		}
 	}
 
@@ -351,11 +354,10 @@ func resourceCCENodePoolV3Read(d *schema.ResourceData, meta interface{}) error {
 			return nil
 		}
 
-		return fmt.Errorf("error retrieving Open Telekom Cloud Node Pool: %s", err)
+		return fmt.Errorf("error retrieving Open Telekom Cloud CCE Node Pool: %s", err)
 	}
 
-	me := &multierror.Error{}
-	me = multierror.Append(me,
+	me := multierror.Append(nil,
 		d.Set("name", s.Metadata.Name),
 		d.Set("flavor", s.Spec.NodeTemplate.Flavor),
 		d.Set("availability_zone", s.Spec.NodeTemplate.Az),
@@ -380,7 +382,7 @@ func resourceCCENodePoolV3Read(d *schema.ResourceData, meta interface{}) error {
 		k8sTags[key] = val
 	}
 	if err := d.Set("k8s_tags", k8sTags); err != nil {
-		return fmt.Errorf("[DEBUG] Error saving k8s_tags to state for Open Telekom Cloud Node Pool (%s): %s", d.Id(), err)
+		return fmt.Errorf("[DEBUG] Error saving k8s_tags to state for Open Telekom Cloud CCE Node Pool (%s): %s", d.Id(), err)
 	}
 
 	var volumes []map[string]interface{}
@@ -392,7 +394,7 @@ func resourceCCENodePoolV3Read(d *schema.ResourceData, meta interface{}) error {
 		volumes = append(volumes, volume)
 	}
 	if err := d.Set("data_volumes", volumes); err != nil {
-		return fmt.Errorf("[DEBUG] Error saving dataVolumes to state for Open Telekom Cloud Node Pool (%s): %s", d.Id(), err)
+		return fmt.Errorf("[DEBUG] Error saving dataVolumes to state for Open Telekom Cloud CCE Node Pool (%s): %s", d.Id(), err)
 	}
 
 	rootVolume := []map[string]interface{}{
@@ -403,11 +405,11 @@ func resourceCCENodePoolV3Read(d *schema.ResourceData, meta interface{}) error {
 		},
 	}
 	if err := d.Set("root_volume", rootVolume); err != nil {
-		return fmt.Errorf("[DEBUG] Error saving rootVolume to state for Open Telekom Cloud Node Pool (%s): %s", d.Id(), err)
+		return fmt.Errorf("[DEBUG] Error saving rootVolume to state for Open Telekom Cloud CCE Node Pool (%s): %s", d.Id(), err)
 	}
 
 	if err := d.Set("status", s.Status.Phase); err != nil {
-		return fmt.Errorf("[DEBUG] Error saving status to state for Open Telekom Cloud Node Pool (%s): %s", d.Id(), err)
+		return fmt.Errorf("[DEBUG] Error saving status to state for Open Telekom Cloud CCE Node Pool (%s): %s", d.Id(), err)
 	}
 
 	return nil
@@ -439,7 +441,7 @@ func resourceCCENodePoolV3Update(d *schema.ResourceData, meta interface{}) error
 	clusterId := d.Get("cluster_id").(string)
 	_, err = nodepools.Update(nodePoolClient, clusterId, d.Id(), updateOpts).Extract()
 	if err != nil {
-		return fmt.Errorf("error updating Open Telekom Cloud Node Pool: %s", err)
+		return fmt.Errorf("error updating Open Telekom Cloud CCE Node Pool: %s", err)
 	}
 	stateConf := &resource.StateChangeConf{
 		Pending:    []string{"Synchronizing"},
@@ -451,7 +453,7 @@ func resourceCCENodePoolV3Update(d *schema.ResourceData, meta interface{}) error
 	}
 	_, err = stateConf.WaitForState()
 	if err != nil {
-		return fmt.Errorf("error creating Open Telekom Cloud Node Pool: %s", err)
+		return fmt.Errorf("error creating Open Telekom Cloud CCE Node Pool: %s", err)
 	}
 
 	return resourceCCENodePoolV3Read(d, meta)
@@ -465,7 +467,7 @@ func resourceCCENodePoolV3Delete(d *schema.ResourceData, meta interface{}) error
 	clusterId := d.Get("cluster_id").(string)
 	err = nodepools.Delete(nodePoolClient, clusterId, d.Id()).ExtractErr()
 	if err != nil {
-		return fmt.Errorf("error deleting Open Telekom Cloud Node Pool: %s", err)
+		return fmt.Errorf("error deleting Open Telekom Cloud CCE Node Pool: %s", err)
 	}
 	stateConf := &resource.StateChangeConf{
 		Pending:      []string{"Deleting"},
@@ -478,7 +480,7 @@ func resourceCCENodePoolV3Delete(d *schema.ResourceData, meta interface{}) error
 
 	_, err = stateConf.WaitForState()
 	if err != nil {
-		return fmt.Errorf("error deleting Open Telekom Cloud Node Pool: %s", err)
+		return fmt.Errorf("error deleting Open Telekom Cloud CCE Node Pool: %s", err)
 	}
 
 	d.SetId("")
@@ -497,13 +499,13 @@ func waitForCceNodePoolActive(cceClient *golangsdk.ServiceClient, clusterId, nod
 
 func waitForCceNodePoolDelete(cceClient *golangsdk.ServiceClient, clusterId, nodePoolId string) resource.StateRefreshFunc {
 	return func() (interface{}, string, error) {
-		log.Printf("[DEBUG] Attempting to delete Open Telekom Cloud Node Pool %s.\n", nodePoolId)
+		log.Printf("[DEBUG] Attempting to delete Open Telekom Cloud CCE Node Pool %s.\n", nodePoolId)
 
 		r, err := nodepools.Get(cceClient, clusterId, nodePoolId).Extract()
 
 		if err != nil {
 			if _, ok := err.(golangsdk.ErrDefault404); ok {
-				log.Printf("[DEBUG] Successfully deleted Open Telekom Cloud Node Pool %s", nodePoolId)
+				log.Printf("[DEBUG] Successfully deleted Open Telekom Cloud CCE Node Pool %s", nodePoolId)
 				return r, "Deleted", nil
 			}
 			return r, "Deleting", err

--- a/opentelekomcloud/resource_opentelekomcloud_cce_node_pool_v3.go
+++ b/opentelekomcloud/resource_opentelekomcloud_cce_node_pool_v3.go
@@ -1,0 +1,206 @@
+package opentelekomcloud
+
+import (
+	"time"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/validation"
+)
+
+func resourceCCENodePoolV3() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceCCENodePoolV3Create,
+		Read:   resourceCCENodePoolV3Read,
+		Update: resourceCCENodePoolV3Update,
+		Delete: resourceCCENodePoolV3Delete,
+
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(20 * time.Minute),
+			Delete: schema.DefaultTimeout(20 * time.Minute),
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"flavor_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"cluster_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"availability_zone": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+				Default:  "random",
+			},
+			"os": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"root_volume": {
+				Type:     schema.TypeList,
+				Required: true,
+				ForceNew: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"size": {
+							Type:     schema.TypeInt,
+							Required: false,
+							Default:  40,
+						},
+						"volume_type": {
+							Type:     schema.TypeString,
+							Required: true,
+							ValidateFunc: validation.StringInSlice([]string{
+								"SATA", "SAS", "SSD",
+							}, true),
+						},
+					}},
+			},
+			"data_volumes": {
+				Type:     schema.TypeList,
+				Required: true,
+				ForceNew: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"size": {
+							Type:         schema.TypeInt,
+							Required:     true,
+							ValidateFunc: validation.IntBetween(10, 32768),
+						},
+						"volume_type": {
+							Type:     schema.TypeString,
+							Required: true,
+							ValidateFunc: validation.StringInSlice([]string{
+								"SATA", "SAS", "SSD", "co-p1", "uh-l1",
+							}, true),
+						},
+						"extend_param": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+					}},
+			},
+			"initial_node_count": {
+				Type:     schema.TypeInt,
+				Required: true,
+			},
+			"k8s_tags": {
+				Type:     schema.TypeMap,
+				Optional: true,
+				ForceNew: true,
+			},
+			"user_tags": {
+				Type:     schema.TypeMap,
+				Optional: true,
+				ForceNew: true,
+			},
+			"taints": {
+				Type:     schema.TypeList,
+				Optional: true,
+				ForceNew: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"key": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"value": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"effect": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+					}},
+			},
+			"key_pair": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				ExactlyOneOf: []string{"password", "key_pair"},
+			},
+			"password": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				Sensitive:    true,
+				ExactlyOneOf: []string{"password", "key_pair"},
+			},
+			"subnet_id": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+			"preinstall": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+				StateFunc: func(v interface{}) string {
+					switch v.(type) {
+					case string:
+						return installScriptHashSum(v.(string))
+					default:
+						return ""
+					}
+				},
+			},
+			"postinstall": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+				StateFunc: func(v interface{}) string {
+					switch v.(type) {
+					case string:
+						return installScriptHashSum(v.(string))
+					default:
+						return ""
+					}
+				},
+			},
+			"scale_enable": {
+				Type:     schema.TypeBool,
+				Optional: true,
+			},
+			"min_node_count": {
+				Type:     schema.TypeInt,
+				Optional: true,
+			},
+			"max_node_count": {
+				Type:     schema.TypeInt,
+				Optional: true,
+			},
+			"scale_down_cooldown_time": {
+				Type:     schema.TypeInt,
+				Optional: true,
+			},
+			"priority": {
+				Type:     schema.TypeInt,
+				Optional: true,
+			},
+		},
+	}
+}
+
+func resourceCCENodePoolV3Create(d *schema.ResourceData, meta interface{}) error {
+	return nil
+}
+func resourceCCENodePoolV3Read(d *schema.ResourceData, meta interface{}) error {
+	return nil
+}
+func resourceCCENodePoolV3Update(d *schema.ResourceData, meta interface{}) error {
+	return nil
+}
+func resourceCCENodePoolV3Delete(d *schema.ResourceData, meta interface{}) error {
+	return nil
+}

--- a/opentelekomcloud/resource_opentelekomcloud_cce_node_pool_v3.go
+++ b/opentelekomcloud/resource_opentelekomcloud_cce_node_pool_v3.go
@@ -518,28 +518,24 @@ func waitForCceNodePoolDelete(cceClient *golangsdk.ServiceClient, clusterId, nod
 }
 
 func recursiveNodePoolCreate(cceClient *golangsdk.ServiceClient, opts nodepools.CreateOptsBuilder, ClusterID string, errCode int) (*nodepools.NodePool, string) {
-	if errCode == 403 {
-		stateCluster := &resource.StateChangeConf{
-			Target:     []string{"Available"},
-			Refresh:    waitForClusterAvailable(cceClient, ClusterID),
-			Timeout:    15 * time.Minute,
-			Delay:      15 * time.Second,
-			MinTimeout: 3 * time.Second,
-		}
-		_, stateErr := stateCluster.WaitForState()
-		if stateErr != nil {
-			log.Printf("[INFO] Cluster Unavailable %s.\n", stateErr)
-		}
-		s, err := nodepools.Create(cceClient, ClusterID, opts).Extract()
-		if err != nil {
-			if _, ok := err.(golangsdk.ErrDefault403); ok {
-				return recursiveNodePoolCreate(cceClient, opts, ClusterID, 403)
-			} else {
-				return s, "fail"
-			}
+	stateCluster := &resource.StateChangeConf{
+		Target:     []string{"Available"},
+		Refresh:    waitForClusterAvailable(cceClient, ClusterID),
+		Timeout:    15 * time.Minute,
+		Delay:      15 * time.Second,
+		MinTimeout: 3 * time.Second,
+	}
+	_, stateErr := stateCluster.WaitForState()
+	if stateErr != nil {
+		log.Printf("[INFO] Cluster Unavailable %s.\n", stateErr)
+	}
+	s, err := nodepools.Create(cceClient, ClusterID, opts).Extract()
+	if err != nil {
+		if _, ok := err.(golangsdk.ErrDefault403); ok {
+			return recursiveNodePoolCreate(cceClient, opts, ClusterID, 403)
 		} else {
-			return s, "success"
+			return s, "fail"
 		}
 	}
-	return nil, "fail"
+	return s, "success"
 }

--- a/opentelekomcloud/resource_opentelekomcloud_cce_node_pool_v3.go
+++ b/opentelekomcloud/resource_opentelekomcloud_cce_node_pool_v3.go
@@ -357,7 +357,7 @@ func resourceCCENodePoolV3Read(d *schema.ResourceData, meta interface{}) error {
 	me := &multierror.Error{}
 	me = multierror.Append(me,
 		d.Set("name", s.Metadata.Name),
-		d.Set("flavor_id", s.Spec.NodeTemplate.Flavor),
+		d.Set("flavor", s.Spec.NodeTemplate.Flavor),
 		d.Set("availability_zone", s.Spec.NodeTemplate.Az),
 		d.Set("os", s.Spec.NodeTemplate.Os),
 		d.Set("key_pair", s.Spec.NodeTemplate.Login.SshKey),

--- a/opentelekomcloud/resource_opentelekomcloud_cce_node_pool_v3.go
+++ b/opentelekomcloud/resource_opentelekomcloud_cce_node_pool_v3.go
@@ -274,7 +274,6 @@ func resourceCCENodePoolV3Create(d *schema.ResourceData, meta interface{}) error
 				Login:       loginSpec,
 				RootVolume:  resourceCCERootVolume(d),
 				DataVolumes: resourceCCEDataVolume(d),
-				K8sTags:     resourceCCENodeK8sTags(d),
 				BillingMode: 0,
 				Count:       1,
 				NodeNicSpec: nodes.NodeNicSpec{
@@ -287,6 +286,7 @@ func resourceCCENodePoolV3Create(d *schema.ResourceData, meta interface{}) error
 					PostInstall: base64PostInstall,
 				},
 				Taints:   resourceCCETaint(d),
+				K8sTags:  resourceCCENodeK8sTags(d),
 				UserTags: resourceCCENodePoolUserTags(d),
 			},
 		},
@@ -367,8 +367,6 @@ func resourceCCENodePoolV3Read(d *schema.ResourceData, meta interface{}) error {
 		d.Set("max_node_count", s.Spec.Autoscaling.MaxNodeCount),
 		d.Set("scale_down_cooldown_time", s.Spec.Autoscaling.ScaleDownCooldownTime),
 		d.Set("priority", s.Spec.Autoscaling.Priority),
-		d.Set("user_tags", s.Spec.NodeTemplate.UserTags),
-		d.Set("taints", s.Spec.NodeTemplate.Taints),
 	)
 	if err := me.ErrorOrNil(); err != nil {
 		return fmt.Errorf("error setting CCE Node Pool attributes (%s): %s", d.Id(), err)

--- a/opentelekomcloud/resource_opentelekomcloud_cce_node_pool_v3.go
+++ b/opentelekomcloud/resource_opentelekomcloud_cce_node_pool_v3.go
@@ -533,9 +533,8 @@ func recursiveNodePoolCreate(cceClient *golangsdk.ServiceClient, opts nodepools.
 	if err != nil {
 		if _, ok := err.(golangsdk.ErrDefault403); ok {
 			return recursiveNodePoolCreate(cceClient, opts, ClusterID, 403)
-		} else {
-			return s, "fail"
 		}
+		return s, "fail"
 	}
 	return s, "success"
 }

--- a/opentelekomcloud/resource_opentelekomcloud_cce_node_pool_v3.go
+++ b/opentelekomcloud/resource_opentelekomcloud_cce_node_pool_v3.go
@@ -61,9 +61,9 @@ func resourceCCENodePoolV3() *schema.Resource {
 				Elem: &schema.Resource{
 					Schema: map[string]*schema.Schema{
 						"size": {
-							Type:     schema.TypeInt,
-							Required: false,
-							Default:  40,
+							Type:         schema.TypeInt,
+							Required:     true,
+							ValidateFunc: validation.IntBetween(10, 32768),
 						},
 						"volume_type": {
 							Type:     schema.TypeString,

--- a/opentelekomcloud/resource_opentelekomcloud_cce_node_pool_v3.go
+++ b/opentelekomcloud/resource_opentelekomcloud_cce_node_pool_v3.go
@@ -285,7 +285,7 @@ func resourceCCENodePoolV3Create(d *schema.ResourceData, meta interface{}) error
 					PreInstall:  base64PreInstall,
 					PostInstall: base64PostInstall,
 				},
-				Taints:   resourceCCETaint(d),
+				Taints:   resourceCCENodeTaints(d),
 				K8sTags:  resourceCCENodeK8sTags(d),
 				UserTags: resourceCCENodePoolUserTags(d),
 			},

--- a/opentelekomcloud/resource_opentelekomcloud_cce_node_pool_v3.go
+++ b/opentelekomcloud/resource_opentelekomcloud_cce_node_pool_v3.go
@@ -1,10 +1,18 @@
 package opentelekomcloud
 
 import (
+	"fmt"
+	"log"
+	"strings"
 	"time"
 
+	"github.com/hashicorp/go-multierror"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
 	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
 	"github.com/hashicorp/terraform-plugin-sdk/helper/validation"
+	"github.com/opentelekomcloud/gophertelekomcloud"
+	"github.com/opentelekomcloud/gophertelekomcloud/openstack/cce/v3/nodepools"
+	"github.com/opentelekomcloud/gophertelekomcloud/openstack/cce/v3/nodes"
 )
 
 func resourceCCENodePoolV3() *schema.Resource {
@@ -24,7 +32,7 @@ func resourceCCENodePoolV3() *schema.Resource {
 				Type:     schema.TypeString,
 				Required: true,
 			},
-			"flavor_id": {
+			"flavor": {
 				Type:     schema.TypeString,
 				Required: true,
 				ForceNew: true,
@@ -64,6 +72,10 @@ func resourceCCENodePoolV3() *schema.Resource {
 								"SATA", "SAS", "SSD",
 							}, true),
 						},
+						"extend_param": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
 					}},
 			},
 			"data_volumes": {
@@ -95,9 +107,13 @@ func resourceCCENodePoolV3() *schema.Resource {
 				Required: true,
 			},
 			"k8s_tags": {
-				Type:     schema.TypeMap,
-				Optional: true,
-				ForceNew: true,
+				Type:         schema.TypeMap,
+				Optional:     true,
+				ForceNew:     true,
+				ValidateFunc: validateK8sTagsMap,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
 			},
 			"user_tags": {
 				Type:     schema.TypeMap,
@@ -188,19 +204,335 @@ func resourceCCENodePoolV3() *schema.Resource {
 				Type:     schema.TypeInt,
 				Optional: true,
 			},
+			"server_group_reference": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+			"status": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
 		},
 	}
 }
 
 func resourceCCENodePoolV3Create(d *schema.ResourceData, meta interface{}) error {
-	return nil
+	config := meta.(*Config)
+	nodePoolClient, err := config.cceV3Client(GetRegion(d, config))
+	if err != nil {
+		return fmt.Errorf("error creating OpenTelekomCloud CCE Node Pool client: %s", err)
+	}
+
+	var base64PreInstall, base64PostInstall string
+	if v, ok := d.GetOk("preinstall"); ok {
+		base64PreInstall = installScriptEncode(v.(string))
+	}
+	if v, ok := d.GetOk("postinstall"); ok {
+		base64PostInstall = installScriptEncode(v.(string))
+	}
+	var loginSpec nodes.LoginSpec
+	if hasFilledOpt(d, "key_pair") {
+		loginSpec = nodes.LoginSpec{SshKey: d.Get("key_pair").(string)}
+	} else if hasFilledOpt(d, "password") {
+		loginSpec = nodes.LoginSpec{
+			UserPassword: nodes.UserPassword{
+				Username: "root",
+				Password: d.Get("password").(string),
+			},
+		}
+	}
+
+	createOpts := nodepools.CreateOpts{
+		Kind:       "NodePool",
+		ApiVersion: "v3",
+		Metadata: nodepools.CreateMetaData{
+			Name: d.Get("name").(string),
+		},
+		Spec: nodepools.CreateSpec{
+			InitialNodeCount: d.Get("initial_node_count").(int),
+			Autoscaling: nodepools.AutoscalingSpec{
+				Enable:                d.Get("scale_enable").(bool),
+				MinNodeCount:          d.Get("min_node_count").(int),
+				MaxNodeCount:          d.Get("max_node_count").(int),
+				ScaleDownCooldownTime: d.Get("scale_down_cooldown_time").(int),
+				Priority:              d.Get("priority").(int),
+			},
+			NodeManagement: nodepools.NodeManagementSpec{
+				ServerGroupReference: d.Get("server_group_reference").(string),
+			},
+			NodeTemplate: nodes.Spec{
+				Flavor:      d.Get("flavor").(string),
+				Az:          d.Get("availability_zone").(string),
+				Os:          d.Get("os").(string),
+				Login:       loginSpec,
+				RootVolume:  resourceCCERootVolume(d),
+				DataVolumes: resourceCCEDataVolume(d),
+				K8sTags:     resourceCCENodeK8sTags(d),
+				BillingMode: 0,
+				Count:       1,
+				NodeNicSpec: nodes.NodeNicSpec{
+					PrimaryNic: nodes.PrimaryNic{
+						SubnetId: d.Get("subnet_id").(string),
+					},
+				},
+				ExtendParam: nodes.ExtendParam{
+					PreInstall:  base64PreInstall,
+					PostInstall: base64PostInstall,
+				},
+				Taints:   resourceCCETaint(d),
+				UserTags: resourceCCENodeTags(d),
+			},
+		},
+	}
+
+	clusterId := d.Get("cluster_id").(string)
+	stateCluster := &resource.StateChangeConf{
+		Target:     []string{"Available"},
+		Refresh:    waitForClusterAvailable(nodePoolClient, clusterId),
+		Timeout:    15 * time.Minute,
+		Delay:      15 * time.Second,
+		MinTimeout: 3 * time.Second,
+	}
+	_, err = stateCluster.WaitForState()
+
+	log.Printf("[DEBUG] Create Options: %#v", createOpts)
+	s, err := nodepools.Create(nodePoolClient, clusterId, createOpts).Extract()
+	if err != nil {
+		if _, ok := err.(golangsdk.ErrDefault403); ok {
+			retryNode, err := recursiveNodePoolCreate(nodePoolClient, createOpts, clusterId, 403)
+			if err == "fail" {
+				return fmt.Errorf("error creating Open Telekom Cloud Node Pool")
+			}
+			s = retryNode
+		} else {
+			return fmt.Errorf("error creating Open Telekom Cloud Node Pool: %s", err)
+		}
+	}
+
+	if len(s.Metadata.Id) == 0 {
+		return fmt.Errorf("error fetching CreateNodePool id")
+	}
+
+	stateConf := &resource.StateChangeConf{
+		Pending:      []string{"Synchronizing"},
+		Target:       []string{""},
+		Refresh:      waitForCceNodePoolActive(nodePoolClient, clusterId, s.Metadata.Id),
+		Timeout:      d.Timeout(schema.TimeoutCreate),
+		Delay:        120 * time.Second,
+		PollInterval: 20 * time.Second,
+	}
+	_, err = stateConf.WaitForState()
+	if err != nil {
+		return fmt.Errorf("error creating Open Telekom Cloud CCE Node Pool: %s", err)
+	}
+
+	d.SetId(s.Metadata.Id)
+	return resourceCCENodePoolV3Read(d, meta)
 }
+
 func resourceCCENodePoolV3Read(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+	nodePoolClient, err := config.cceV3Client(GetRegion(d, config))
+	if err != nil {
+		return fmt.Errorf("error creating Open Telekom Cloud CCE Node Pool client: %s", err)
+	}
+	clusterId := d.Get("cluster_id").(string)
+	s, err := nodepools.Get(nodePoolClient, clusterId, d.Id()).Extract()
+	if err != nil {
+		if _, ok := err.(golangsdk.ErrDefault404); ok {
+			d.SetId("")
+			return nil
+		}
+
+		return fmt.Errorf("error retrieving Open Telekom Cloud Node Pool: %s", err)
+	}
+
+	me := &multierror.Error{}
+	me = multierror.Append(me,
+		d.Set("name", s.Metadata.Name),
+		d.Set("flavor_id", s.Spec.NodeTemplate.Flavor),
+		d.Set("availability_zone", s.Spec.NodeTemplate.Az),
+		d.Set("os", s.Spec.NodeTemplate.Os),
+		d.Set("key_pair", s.Spec.NodeTemplate.Login.SshKey),
+		d.Set("initial_node_count", s.Spec.InitialNodeCount),
+		d.Set("scale_enable", s.Spec.Autoscaling.Enable),
+		d.Set("min_node_count", s.Spec.Autoscaling.MinNodeCount),
+		d.Set("max_node_count", s.Spec.Autoscaling.MaxNodeCount),
+		d.Set("scale_down_cooldown_time", s.Spec.Autoscaling.ScaleDownCooldownTime),
+		d.Set("priority", s.Spec.Autoscaling.Priority),
+		d.Set("user_tags", s.Spec.NodeTemplate.UserTags),
+		d.Set("taints", s.Spec.NodeTemplate.Taints),
+	)
+	if err := me.ErrorOrNil(); err != nil {
+		return fmt.Errorf("error setting CCE Node Pool attributes (%s): %s", d.Id(), err)
+	}
+
+	k8sTags := map[string]string{}
+	for key, val := range s.Spec.NodeTemplate.K8sTags {
+		if strings.Contains(key, "cce.cloud.com") {
+			continue
+		}
+		k8sTags[key] = val
+	}
+	if err := d.Set("k8s_tags", k8sTags); err != nil {
+		return fmt.Errorf("[DEBUG] Error saving k8s_tags to state for Open Telekom Cloud Node Pool (%s): %s", d.Id(), err)
+	}
+
+	var volumes []map[string]interface{}
+	for _, pairObject := range s.Spec.NodeTemplate.DataVolumes {
+		volume := make(map[string]interface{})
+		volume["size"] = pairObject.Size
+		volume["volumetype"] = pairObject.VolumeType
+		volume["extend_param"] = pairObject.ExtendParam
+		volumes = append(volumes, volume)
+	}
+	if err := d.Set("data_volumes", volumes); err != nil {
+		return fmt.Errorf("[DEBUG] Error saving dataVolumes to state for Open Telekom Cloud Node Pool (%s): %s", d.Id(), err)
+	}
+
+	rootVolume := []map[string]interface{}{
+		{
+			"size":         s.Spec.NodeTemplate.RootVolume.Size,
+			"volumetype":   s.Spec.NodeTemplate.RootVolume.VolumeType,
+			"extend_param": s.Spec.NodeTemplate.RootVolume.ExtendParam,
+		},
+	}
+	if err := d.Set("root_volume", rootVolume); err != nil {
+		return fmt.Errorf("[DEBUG] Error saving rootVolume to state for Open Telekom Cloud Node Pool (%s): %s", d.Id(), err)
+	}
+
+	if err := d.Set("status", s.Status.Phase); err != nil {
+		return fmt.Errorf("[DEBUG] Error saving status to state for Open Telekom Cloud Node Pool (%s): %s", d.Id(), err)
+	}
+
 	return nil
 }
+
 func resourceCCENodePoolV3Update(d *schema.ResourceData, meta interface{}) error {
-	return nil
+	config := meta.(*Config)
+	nodePoolClient, err := config.cceV3Client(GetRegion(d, config))
+	if err != nil {
+		return fmt.Errorf("error creating Open Telekom Cloud CCE client: %s", err)
+	}
+	updateOpts := nodepools.UpdateOpts{
+		Kind:       "NodePool",
+		ApiVersion: "v3",
+		Metadata: nodepools.UpdateMetaData{
+			Name: d.Get("name").(string),
+		},
+		Spec: nodepools.UpdateSpec{
+			InitialNodeCount: d.Get("initial_node_count").(int),
+			Autoscaling: nodepools.AutoscalingSpec{
+				Enable:                d.Get("scale_enable").(bool),
+				MinNodeCount:          d.Get("min_node_count").(int),
+				MaxNodeCount:          d.Get("max_node_count").(int),
+				ScaleDownCooldownTime: d.Get("scale_down_cooldown_time").(int),
+				Priority:              d.Get("priority").(int),
+			},
+		},
+	}
+	clusterId := d.Get("cluster_id").(string)
+	_, err = nodepools.Update(nodePoolClient, clusterId, d.Id(), updateOpts).Extract()
+	if err != nil {
+		return fmt.Errorf("error updating Open Telekom Cloud Node Pool: %s", err)
+	}
+	stateConf := &resource.StateChangeConf{
+		Pending:    []string{"Synchronizing"},
+		Target:     []string{""},
+		Refresh:    waitForCceNodePoolActive(nodePoolClient, clusterId, d.Id()),
+		Timeout:    d.Timeout(schema.TimeoutCreate),
+		Delay:      15 * time.Second,
+		MinTimeout: 5 * time.Second,
+	}
+	_, err = stateConf.WaitForState()
+	if err != nil {
+		return fmt.Errorf("error creating Open Telekom Cloud Node Pool: %s", err)
+	}
+
+	return resourceCCENodePoolV3Read(d, meta)
 }
 func resourceCCENodePoolV3Delete(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+	nodePoolClient, err := config.cceV3Client(GetRegion(d, config))
+	if err != nil {
+		return fmt.Errorf("error creating Open Telekom Cloud CCE client: %s", err)
+	}
+	clusterId := d.Get("cluster_id").(string)
+	err = nodepools.Delete(nodePoolClient, clusterId, d.Id()).ExtractErr()
+	if err != nil {
+		return fmt.Errorf("error deleting Open Telekom Cloud Node Pool: %s", err)
+	}
+	stateConf := &resource.StateChangeConf{
+		Pending:      []string{"Deleting"},
+		Target:       []string{"Deleted"},
+		Refresh:      waitForCceNodePoolDelete(nodePoolClient, clusterId, d.Id()),
+		Timeout:      d.Timeout(schema.TimeoutDelete),
+		Delay:        60 * time.Second,
+		PollInterval: 20 * time.Second,
+	}
+
+	_, err = stateConf.WaitForState()
+	if err != nil {
+		return fmt.Errorf("error deleting Open Telekom Cloud Node Pool: %s", err)
+	}
+
+	d.SetId("")
 	return nil
+}
+
+func waitForCceNodePoolActive(cceClient *golangsdk.ServiceClient, clusterId, nodePoolId string) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		n, err := nodepools.Get(cceClient, clusterId, nodePoolId).Extract()
+		if err != nil {
+			return nil, "", err
+		}
+		return n, n.Status.Phase, nil
+	}
+}
+
+func waitForCceNodePoolDelete(cceClient *golangsdk.ServiceClient, clusterId, nodePoolId string) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		log.Printf("[DEBUG] Attempting to delete Open Telekom Cloud Node Pool %s.\n", nodePoolId)
+
+		r, err := nodepools.Get(cceClient, clusterId, nodePoolId).Extract()
+
+		if err != nil {
+			if _, ok := err.(golangsdk.ErrDefault404); ok {
+				log.Printf("[DEBUG] Successfully deleted Open Telekom Cloud Node Pool %s", nodePoolId)
+				return r, "Deleted", nil
+			}
+			return r, "Deleting", err
+		}
+
+		log.Printf("[DEBUG] Open Telekom Cloud Node Pool %s still available.\n", nodePoolId)
+		return r, r.Status.Phase, nil
+	}
+}
+
+func recursiveNodePoolCreate(cceClient *golangsdk.ServiceClient, opts nodepools.CreateOptsBuilder, ClusterID string, errCode int) (*nodepools.NodePool, string) {
+	if errCode == 403 {
+		stateCluster := &resource.StateChangeConf{
+			Target:     []string{"Available"},
+			Refresh:    waitForClusterAvailable(cceClient, ClusterID),
+			Timeout:    15 * time.Minute,
+			Delay:      15 * time.Second,
+			MinTimeout: 3 * time.Second,
+		}
+		_, stateErr := stateCluster.WaitForState()
+		if stateErr != nil {
+			log.Printf("[INFO] Cluster Unavailable %s.\n", stateErr)
+		}
+		s, err := nodepools.Create(cceClient, ClusterID, opts).Extract()
+		if err != nil {
+			if _, ok := err.(golangsdk.ErrDefault403); ok {
+				return recursiveNodePoolCreate(cceClient, opts, ClusterID, 403)
+			} else {
+				return s, "fail"
+			}
+		} else {
+			return s, "success"
+		}
+	}
+	return nil, "fail"
 }

--- a/opentelekomcloud/resource_opentelekomcloud_cce_node_pool_v3.go
+++ b/opentelekomcloud/resource_opentelekomcloud_cce_node_pool_v3.go
@@ -13,6 +13,7 @@ import (
 	"github.com/opentelekomcloud/gophertelekomcloud"
 	"github.com/opentelekomcloud/gophertelekomcloud/openstack/cce/v3/nodepools"
 	"github.com/opentelekomcloud/gophertelekomcloud/openstack/cce/v3/nodes"
+	"github.com/opentelekomcloud/gophertelekomcloud/openstack/common/tags"
 )
 
 func resourceCCENodePoolV3() *schema.Resource {
@@ -65,7 +66,7 @@ func resourceCCENodePoolV3() *schema.Resource {
 							Required:     true,
 							ValidateFunc: validation.IntBetween(10, 32768),
 						},
-						"volume_type": {
+						"volumetype": {
 							Type:     schema.TypeString,
 							Required: true,
 							ValidateFunc: validation.StringInSlice([]string{
@@ -89,7 +90,7 @@ func resourceCCENodePoolV3() *schema.Resource {
 							Required:     true,
 							ValidateFunc: validation.IntBetween(10, 32768),
 						},
-						"volume_type": {
+						"volumetype": {
 							Type:     schema.TypeString,
 							Required: true,
 							ValidateFunc: validation.StringInSlice([]string{
@@ -217,6 +218,11 @@ func resourceCCENodePoolV3() *schema.Resource {
 	}
 }
 
+func resourceCCENodePoolUserTags(d *schema.ResourceData) []tags.ResourceTag {
+	tagRaw := d.Get("user_tags").(map[string]interface{})
+	return expandResourceTags(tagRaw)
+}
+
 func resourceCCENodePoolV3Create(d *schema.ResourceData, meta interface{}) error {
 	config := meta.(*Config)
 	nodePoolClient, err := config.cceV3Client(GetRegion(d, config))
@@ -281,7 +287,7 @@ func resourceCCENodePoolV3Create(d *schema.ResourceData, meta interface{}) error
 					PostInstall: base64PostInstall,
 				},
 				Taints:   resourceCCETaint(d),
-				UserTags: resourceCCENodeTags(d),
+				UserTags: resourceCCENodePoolUserTags(d),
 			},
 		},
 	}

--- a/opentelekomcloud/resource_opentelekomcloud_cce_node_pool_v3.go
+++ b/opentelekomcloud/resource_opentelekomcloud_cce_node_pool_v3.go
@@ -88,7 +88,7 @@ func resourceCCENodePoolV3() *schema.Resource {
 						"size": {
 							Type:         schema.TypeInt,
 							Required:     true,
-							ValidateFunc: validation.IntBetween(10, 32768),
+							ValidateFunc: validation.IntBetween(100, 32768),
 						},
 						"volumetype": {
 							Type:     schema.TypeString,

--- a/opentelekomcloud/resource_opentelekomcloud_cce_node_pool_v3.go
+++ b/opentelekomcloud/resource_opentelekomcloud_cce_node_pool_v3.go
@@ -243,7 +243,8 @@ func resourceCCENodePoolV3Create(d *schema.ResourceData, meta interface{}) error
 	var loginSpec nodes.LoginSpec
 	if hasFilledOpt(d, "key_pair") {
 		loginSpec = nodes.LoginSpec{SshKey: d.Get("key_pair").(string)}
-	} else if hasFilledOpt(d, "password") {
+	}
+	if hasFilledOpt(d, "password") {
 		loginSpec = nodes.LoginSpec{
 			UserPassword: nodes.UserPassword{
 				Username: "root",

--- a/opentelekomcloud/resource_opentelekomcloud_cce_node_pool_v3_test.go
+++ b/opentelekomcloud/resource_opentelekomcloud_cce_node_pool_v3_test.go
@@ -105,7 +105,7 @@ func testAccCheckCCENodePoolV3Exists(n string, cluster string, nodepool *nodepoo
 
 var testAccCCENodePoolV3_basic = fmt.Sprintf(`
 resource "opentelekomcloud_cce_cluster_v3" "cluster" {
-  name         = "opentelekomcloud-cce"
+  name         = "opentelekomcloud-cce-np"
   cluster_type = "VirtualMachine"
   flavor_id    = "cce.s1.small"
   vpc_id       = "%s"
@@ -119,7 +119,7 @@ resource "opentelekomcloud_cce_node_pool_v3" "node_pool" {
   cluster_id         = opentelekomcloud_cce_cluster_v3.cluster.id
   name               = "opentelekomcloud-cce-node-pool"
   os                 = "EulerOS 2.5"
-  flavor             = "s6.large.2"
+  flavor             = "s2.xlarge.2"
   initial_node_count = 1
   availability_zone  = "%s"
   key_pair           = "%s"
@@ -135,14 +135,14 @@ resource "opentelekomcloud_cce_node_pool_v3" "node_pool" {
     volumetype = "SSD"
   }
   data_volumes {
-    size       = 50
+    size       = 100
     volumetype = "SSD"
   }
 }`, OS_VPC_ID, OS_NETWORK_ID, OS_AVAILABILITY_ZONE, OS_KEYPAIR_NAME)
 
 var testAccCCENodePoolV3_update = fmt.Sprintf(`
 resource "opentelekomcloud_cce_cluster_v3" "cluster" {
-  name         = "opentelekomcloud-cce"
+  name         = "opentelekomcloud-cce-np"
   cluster_type = "VirtualMachine"
   flavor_id    = "cce.s1.small"
   vpc_id       = "%s"
@@ -156,7 +156,7 @@ resource "opentelekomcloud_cce_node_pool_v3" "node_pool" {
   cluster_id         = opentelekomcloud_cce_cluster_v3.cluster.id
   name               = "opentelekomcloud-cce-node-pool"
   os                 = "EulerOS 2.5"
-  flavor             = "s6.large.2"
+  flavor             = "s2.xlarge.2"
   initial_node_count = 2
   availability_zone  = "%s"
   key_pair           = "%s"
@@ -172,7 +172,7 @@ resource "opentelekomcloud_cce_node_pool_v3" "node_pool" {
     volumetype = "SSD"
   }
   data_volumes {
-    size       = 50
+    size       = 100
     volumetype = "SSD"
   }
 }`, OS_VPC_ID, OS_NETWORK_ID, OS_AVAILABILITY_ZONE, OS_KEYPAIR_NAME)

--- a/opentelekomcloud/resource_opentelekomcloud_cce_node_pool_v3_test.go
+++ b/opentelekomcloud/resource_opentelekomcloud_cce_node_pool_v3_test.go
@@ -141,6 +141,17 @@ resource "opentelekomcloud_cce_node_pool_v3" "node_pool" {
 }`, OS_VPC_ID, OS_NETWORK_ID, OS_AVAILABILITY_ZONE, OS_KEYPAIR_NAME)
 
 var testAccCCENodePoolV3_update = fmt.Sprintf(`
+resource "opentelekomcloud_cce_cluster_v3" "cluster" {
+  name         = "opentelekomcloud-cce"
+  cluster_type = "VirtualMachine"
+  flavor_id    = "cce.s1.small"
+  vpc_id       = "%s"
+  subnet_id    = "%s"
+
+  container_network_type = "overlay_l2"
+  authentication_mode    = "rbac"
+}
+
 resource "opentelekomcloud_cce_node_pool_v3" "node_pool" {
   cluster_id         = opentelekomcloud_cce_cluster_v3.cluster.id
   name               = "opentelekomcloud-cce-node-pool"
@@ -164,4 +175,4 @@ resource "opentelekomcloud_cce_node_pool_v3" "node_pool" {
     size       = 50
     volumetype = "SSD"
   }
-}`, OS_AVAILABILITY_ZONE, OS_KEYPAIR_NAME)
+}`, OS_VPC_ID, OS_NETWORK_ID, OS_AVAILABILITY_ZONE, OS_KEYPAIR_NAME)

--- a/opentelekomcloud/resource_opentelekomcloud_cce_node_pool_v3_test.go
+++ b/opentelekomcloud/resource_opentelekomcloud_cce_node_pool_v3_test.go
@@ -23,14 +23,14 @@ func TestAccCCENodePoolsV3_basic(t *testing.T) {
 				Check: resource.ComposeTestCheckFunc(
 					testAccCheckCCENodePoolV3Exists("opentelekomcloud_cce_node_pool_v3.node_pool", "opentelekomcloud_cce_cluster_v3.cluster", &nodePool),
 					resource.TestCheckResourceAttr("opentelekomcloud_cce_node_pool_v3.node_pool", "name", "opentelekomcloud-cce-node-pool"),
-					resource.TestCheckResourceAttr("opentelekomcloud_cce_node_pool_v3.node_pool", "flavor_id", "s6.large.2"),
+					resource.TestCheckResourceAttr("opentelekomcloud_cce_node_pool_v3.node_pool", "flavor", "s2.xlarge.2"),
 					resource.TestCheckResourceAttr("opentelekomcloud_cce_node_pool_v3.node_pool", "os", "EulerOS 2.5"),
 				),
 			},
 			{
 				Config: testAccCCENodePoolV3_update,
 				Check: resource.ComposeTestCheckFunc(
-					resource.TestCheckResourceAttr("opentelekomcloud_cce_node_v3.node_pool", "initial_node_count", "2"),
+					resource.TestCheckResourceAttr("opentelekomcloud_cce_node_pool_v3.node_pool", "initial_node_count", "2"),
 				),
 			},
 		},

--- a/opentelekomcloud/resource_opentelekomcloud_cce_node_pool_v3_test.go
+++ b/opentelekomcloud/resource_opentelekomcloud_cce_node_pool_v3_test.go
@@ -132,13 +132,12 @@ resource "opentelekomcloud_cce_node_pool_v3" "node_pool" {
 
   root_volume {
     size       = 40
-    volumetype = "SSD"
+    volume_type = "SSD"
   }
   data_volumes {
     size       = 50
-    volumetype = "SSD"
+    volume_type = "SSD"
   }
-}
 }`, OS_VPC_ID, OS_NETWORK_ID, OS_AVAILABILITY_ZONE, OS_KEYPAIR_NAME)
 
 var testAccCCENodePoolV3_update = fmt.Sprintf(`
@@ -159,11 +158,10 @@ resource "opentelekomcloud_cce_node_pool_v3" "node_pool" {
 
   root_volume {
     size       = 40
-    volumetype = "SSD"
+    volume_type = "SSD"
   }
   data_volumes {
     size       = 50
-    volumetype = "SSD"
+    volume_type = "SSD"
   }
-}
 }`, OS_AVAILABILITY_ZONE, OS_KEYPAIR_NAME)

--- a/opentelekomcloud/resource_opentelekomcloud_cce_node_pool_v3_test.go
+++ b/opentelekomcloud/resource_opentelekomcloud_cce_node_pool_v3_test.go
@@ -132,11 +132,11 @@ resource "opentelekomcloud_cce_node_pool_v3" "node_pool" {
 
   root_volume {
     size       = 40
-    volume_type = "SSD"
+    volumetype = "SSD"
   }
   data_volumes {
     size       = 50
-    volume_type = "SSD"
+    volumetype = "SSD"
   }
 }`, OS_VPC_ID, OS_NETWORK_ID, OS_AVAILABILITY_ZONE, OS_KEYPAIR_NAME)
 
@@ -158,10 +158,10 @@ resource "opentelekomcloud_cce_node_pool_v3" "node_pool" {
 
   root_volume {
     size       = 40
-    volume_type = "SSD"
+    volumetype = "SSD"
   }
   data_volumes {
     size       = 50
-    volume_type = "SSD"
+    volumetype = "SSD"
   }
 }`, OS_AVAILABILITY_ZONE, OS_KEYPAIR_NAME)

--- a/opentelekomcloud/resource_opentelekomcloud_cce_node_pool_v3_test.go
+++ b/opentelekomcloud/resource_opentelekomcloud_cce_node_pool_v3_test.go
@@ -1,0 +1,1 @@
+package opentelekomcloud

--- a/opentelekomcloud/resource_opentelekomcloud_cce_node_pool_v3_test.go
+++ b/opentelekomcloud/resource_opentelekomcloud_cce_node_pool_v3_test.go
@@ -1,1 +1,169 @@
 package opentelekomcloud
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/terraform"
+
+	"github.com/opentelekomcloud/gophertelekomcloud/openstack/cce/v3/nodepools"
+)
+
+func TestAccCCENodePoolsV3_basic(t *testing.T) {
+	var nodePool nodepools.NodePool
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccCCEKeyPairPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckCCENodePoolV3Destroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccCCENodePoolV3_basic,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckCCENodePoolV3Exists("opentelekomcloud_cce_node_pool_v3.node_pool", "opentelekomcloud_cce_cluster_v3.cluster", &nodePool),
+					resource.TestCheckResourceAttr("opentelekomcloud_cce_node_pool_v3.node_pool", "name", "opentelekomcloud-cce-node-pool"),
+					resource.TestCheckResourceAttr("opentelekomcloud_cce_node_pool_v3.node_pool", "flavor_id", "s6.large.2"),
+					resource.TestCheckResourceAttr("opentelekomcloud_cce_node_pool_v3.node_pool", "os", "EulerOS 2.5"),
+				),
+			},
+			{
+				Config: testAccCCENodePoolV3_update,
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("opentelekomcloud_cce_node_v3.node_pool", "initial_node_count", "2"),
+				),
+			},
+		},
+	})
+}
+
+func testAccCheckCCENodePoolV3Destroy(s *terraform.State) error {
+	config := testAccProvider.Meta().(*Config)
+	cceClient, err := config.cceV3Client(OS_REGION_NAME)
+	if err != nil {
+		return fmt.Errorf("error creating OpenTelekomCloud CCE client: %s", err)
+	}
+
+	var clusterId string
+
+	for _, rs := range s.RootModule().Resources {
+		if rs.Type == "opentelekomcloud_cce_cluster_v3" {
+			clusterId = rs.Primary.ID
+		}
+
+		if rs.Type != "opentelekomcloud_cce_node_pool_v3" {
+			continue
+		}
+
+		_, err := nodepools.Get(cceClient, clusterId, rs.Primary.ID).Extract()
+		if err == nil {
+			return fmt.Errorf("node pool still exists")
+		}
+	}
+
+	return nil
+}
+
+func testAccCheckCCENodePoolV3Exists(n string, cluster string, nodepool *nodepools.NodePool) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		rs, ok := s.RootModule().Resources[n]
+		if !ok {
+			return fmt.Errorf("not found: %s", n)
+		}
+		c, ok := s.RootModule().Resources[cluster]
+		if !ok {
+			return fmt.Errorf("cluster not found: %s", c)
+		}
+
+		if rs.Primary.ID == "" {
+			return fmt.Errorf("no ID is set")
+		}
+		if c.Primary.ID == "" {
+			return fmt.Errorf("cluster id is not set")
+		}
+
+		config := testAccProvider.Meta().(*Config)
+		cceClient, err := config.cceV3Client(OS_REGION_NAME)
+		if err != nil {
+			return fmt.Errorf("error creating OpenTelekomCloud CCE client: %s", err)
+		}
+
+		found, err := nodepools.Get(cceClient, c.Primary.ID, rs.Primary.ID).Extract()
+		if err != nil {
+			return err
+		}
+
+		if found.Metadata.Id != rs.Primary.ID {
+			return fmt.Errorf("node pool not found")
+		}
+
+		*nodepool = *found
+
+		return nil
+	}
+}
+
+var testAccCCENodePoolV3_basic = fmt.Sprintf(`
+resource "opentelekomcloud_cce_cluster_v3" "cluster" {
+  name         = "opentelekomcloud-cce"
+  cluster_type = "VirtualMachine"
+  flavor_id    = "cce.s1.small"
+  vpc_id       = "%s"
+  subnet_id    = "%s"
+
+  container_network_type = "overlay_l2"
+  authentication_mode    = "rbac"
+}
+
+resource "opentelekomcloud_cce_node_pool_v3" "node_pool" {
+  cluster_id         = opentelekomcloud_cce_cluster_v3.cluster.id
+  name               = "opentelekomcloud-cce-node-pool"
+  os                 = "EulerOS 2.5"
+  flavor             = "s6.large.2"
+  initial_node_count = 1
+  availability_zone  = "%s"
+  key_pair           = "%s"
+
+  scale_enable             = false
+  min_node_count           = 0
+  max_node_count           = 0
+  scale_down_cooldown_time = 0
+  priority                 = 0
+
+  root_volume {
+    size       = 40
+    volumetype = "SSD"
+  }
+  data_volumes {
+    size       = 50
+    volumetype = "SSD"
+  }
+}
+}`, OS_VPC_ID, OS_NETWORK_ID, OS_AVAILABILITY_ZONE, OS_KEYPAIR_NAME)
+
+var testAccCCENodePoolV3_update = fmt.Sprintf(`
+resource "opentelekomcloud_cce_node_pool_v3" "node_pool" {
+  cluster_id         = opentelekomcloud_cce_cluster_v3.cluster.id
+  name               = "opentelekomcloud-cce-node-pool"
+  os                 = "EulerOS 2.5"
+  flavor             = "s6.large.2"
+  initial_node_count = 2
+  availability_zone  = "%s"
+  key_pair           = "%s"
+
+  scale_enable             = true
+  min_node_count           = 2
+  max_node_count           = 9
+  scale_down_cooldown_time = 100
+  priority                 = 1
+
+  root_volume {
+    size       = 40
+    volumetype = "SSD"
+  }
+  data_volumes {
+    size       = 50
+    volumetype = "SSD"
+  }
+}
+}`, OS_AVAILABILITY_ZONE, OS_KEYPAIR_NAME)

--- a/opentelekomcloud/resource_opentelekomcloud_cce_node_v3.go
+++ b/opentelekomcloud/resource_opentelekomcloud_cce_node_v3.go
@@ -300,7 +300,7 @@ func resourceCCENodeTags(d *schema.ResourceData) []tags.ResourceTag {
 	return expandResourceTags(tagRaw)
 }
 
-func resourceCCETaint(d *schema.ResourceData) []nodes.TaintSpec {
+func resourceCCENodeTaints(d *schema.ResourceData) []nodes.TaintSpec {
 	taintRaw := d.Get("taints").([]interface{})
 	taints := make([]nodes.TaintSpec, len(taintRaw))
 	for i, raw := range taintRaw {

--- a/opentelekomcloud/resource_opentelekomcloud_cce_node_v3.go
+++ b/opentelekomcloud/resource_opentelekomcloud_cce_node_v3.go
@@ -300,6 +300,20 @@ func resourceCCENodeTags(d *schema.ResourceData) []tags.ResourceTag {
 	return expandResourceTags(tagRaw)
 }
 
+func resourceCCETaint(d *schema.ResourceData) []nodes.TaintSpec {
+	taintRaw := d.Get("taints").([]interface{})
+	taints := make([]nodes.TaintSpec, len(taintRaw))
+	for i, raw := range taintRaw {
+		rawMap := raw.(map[string]interface{})
+		taints[i] = nodes.TaintSpec{
+			Key:    rawMap["key"].(string),
+			Value:  rawMap["value"].(string),
+			Effect: rawMap["effect"].(string),
+		}
+	}
+	return taints
+}
+
 func resourceCCENodeK8sTags(d *schema.ResourceData) map[string]string {
 	m := make(map[string]string)
 	for key, val := range d.Get("k8s_tags").(map[string]interface{}) {

